TickSpec/ServiceProvider.fs

```diff
@@ -3,21 +3,94 @@
+    let rec resolveInstance (t:Type) (typeStack: Type list) =
+        let alreadyRequested =
+            typeStack
+            |> List.tryFind (fun x -> x = t)
+
+        match alreadyRequested with
+        | Some _ -> raise (InvalidOperationException(sprintf "Circular dependency found when resolving type %O" t))
+        | None -> ()
+
+        match t with
+        | t when t = typeof<IInstanceProvider> -> self :> obj
+        | t ->
+            match instances.TryGetValue t with
+            | true, instance -> instance
+            | false, _ ->
+                let instance = createInstance t typeStack
+                instances.Add(t, instance)
+                instance
+
+    /// Creates an instance if there was none for a speicified type (and remembering the stack of types being resolved)
+    and createInstance (t:Type) (typeStack: Type list) =
+        let constructors =
+            t.GetConstructors()
+            |> List.ofArray
+
+        let (_, widestConstructors) =
+            constructors
+            |> List.map (fun x -> (x.GetParameters().Length, x))
+            |> List.fold (fun (widest, constructorList) (parameterCount, c) ->
+                if parameterCount > widest then
+                    (parameterCount, [ c ])
+                elif parameterCount = widest then
+                    (parameterCount, c :: constructorList)
+                else
+                    (widest, constructorList)
+            ) (-1, [])
+
+        let createObject (c: ConstructorInfo) =
+            let resolveArgument (arg: ParameterInfo) =
+                resolveInstance arg.ParameterType (t::typeStack)
+
+            c.GetParameters()
+            |> Array.map resolveArgument
+            |> c.Invoke
+
+        match widestConstructors with
+        | [] -> raise (InvalidOperationException(sprintf "The type does not have any public constructor: %O" t))
+        | [ c ] -> c |> createObject
+        | _ -> raise (InvalidOperationException(sprintf "Cannot decide which constructor to use. The type has multiple constructors with the same maximum number of parameters: %O" t))
+
     /// Gets type instance for specified type
-    [<DebuggerStepThrough>]
     let getInstance (t:Type) =
-        match instances.TryGetValue t with
-        | true, instance -> instance
-        | false, _ ->
-            let instance = Activator.CreateInstance t
-            instances.Add(t,instance)
-            instance
-    interface System.IServiceProvider with
+        resolveInstance t []
+
+    interface IInstanceProvider with
         [<DebuggerStepThrough>]
-        member this.GetService(t:Type) =
+        member this.Resolve (t: Type) =
             getInstance t
+
+        member this.RegisterInstance (t: Type) (instance: obj) =
+            instances.[t] <- instance
+
+    interface IDisposable with
+        member this.Dispose() =
+            instances.Values
+            |> Seq.choose (fun x ->

```

Comment:
I would merge the iter and the choose

Or 

```
for i in instances.Values do
    match i with
    | :? Disposable as disposable -> disposable.Dispose()
    | _ -> ()  
```

Reply:
Also function could be use instead of fun + match